runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -1,14 +1,269 @@
+                            log.info("Snapshot sync was going on during multi table read.  Rerun the checks and " +
+                                "re-trigger multiTable read");
+                            throw new RetryNeededException();
+                        }
+
+                        // Snapshot sync was not in progress at the client's read timestamp.  Invoke the client callback
+                        // to set the baseline at this timestamp.
+                        clientListener.mergeTableOnInitialSubscription(multiTableReadTimestamp);
+
+                        // Return this timestamp.  The client's listener must be subscribed at this timestamp.
+                        return multiTableReadTimestamp;
+                    } catch (Exception e) {
+                        log.error("Error while attempting to get a valid subscription timestamp.  Retrying.", e);
+                        throw new RetryNeededException();
+                    }
+                }).setOptions(x -> {
+                    x.setBackoffDuration(Duration.ofMillis(backoff_retry_duration_ms));
+                    x.setMaxRetryThreshold(Duration.ofMinutes(max_duration_threshold_mins));
+                }).run();
+            } catch (InterruptedException e) {
+                log.error("Failed to get a valid subscription timestamp.", e);
+                clientListener.onError(e);
+            }
+            return null;
+        }
+
+        /**
+         * Gets the latest token from the sequencer and builds a timestamp from it
+         * @return Timestamp
+         */
+        private CorfuStoreMetadata.Timestamp getTimestamp() {
+            Token token = corfuStore.getRuntime().getSequencerView().query().getToken();
+            return CorfuStoreMetadata.Timestamp.newBuilder()
+                .setEpoch(token.getEpoch())
+                .setSequence(token.getSequence())
+                .build();
+        }
+
+        /**
+         * Checks if snapshot sync is ongoing at the given timestamp.
+         * @param timestamp
+         * @return
+         */
+        private boolean checkSnapshotSyncOngoing(CorfuStoreMetadata.Timestamp timestamp) {
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE, IsolationLevel.snapshot(timestamp))) {
+                List<CorfuStoreEntry<ReplicationStatusKey, ReplicationStatusVal, Message>> entries =
+                    txn.executeQuery(REPLICATION_STATUS_TABLE, p -> true);
+                // In LR V1, it is a valid assumption that the size of replication status table will be 1 as there is
+                // only 1 remote cluster.  This implementation will change in LR V2
+                Preconditions.checkState(entries.size() == 1);
+
+                CorfuStoreEntry<ReplicationStatusKey, ReplicationStatusVal, Message> entry = entries.get(0);
+                txn.commit();
+                return entry.getPayload().getDataConsistent() ? false : true;
```

Comment:
1. I think its fine, but just to confirm, payload and dataConsistent will never be null, right?
2. Can just return `!entry.getPayload().getDataConsistent() `